### ISSUE TITLE
R_MOS_ADDR16_XX relocations has wrong offset

### ISSUE THREAD

**Issue by laoo** [johnwbyrd/llvm-mos#2]
Assembling the snippet:
```
_start:
    lda #mos16lo(abs)
    lda abs
    lda #mos16hi(abs)
    jsr abs
```

produces such disassembly:

```
00000000 <_start>:
       0: 00                            brk
       1: 00                            brk
       2: a5 00                         lda     0
       4: 00                            brk
       5: 00                            brk
       6: 20 00 00                      jsr     0
```

And such relocation table:

```
RELOCATION RECORDS FOR [.text]:
OFFSET   TYPE                     VALUE
00000000 R_MOS_ADDR16_LO          abs
00000003 R_MOS_ADDR8              abs
00000004 R_MOS_ADDR16_HI          abs
00000007 R_MOS_ADDR16             abs
```

While `R_MOS_ADDR8` and `R_MOS_ADDR16` are fine, the `R_MOS_ADDR16_LO` and `R_MOS_ADDR16_HI` offsets are off by one byte. Additionally the opcodes are zeroed. I'm not sure which is the cause and which the effect.

**Comment by laoo** [comment]
Hi. I've read the sources and found that fixups arent performed on `MOSMCExpr` expressions. `Addr16_Low` and `Addr16_High` are involved in expressions of that type hence no fixups for them. I don't know now how to resolve it but temporarily for my own use cases I've made a patch to exclude these to cases. Probably it won't work for some more complicated scenarios but trivial one seems to be working:

```diff
diff --git a/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.cpp b/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.cpp
index 9117064c162..0d7c526b131 100644
--- a/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.cpp
+++ b/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.cpp
@@ -15,6 +15,7 @@
 
 #include "MCTargetDesc/MOSMCExpr.h"
 #include "MCTargetDesc/MOSMCTargetDesc.h"
+#include "MCTargetDesc/MOSFixupKinds.h"
 
 #include "llvm/ADT/APFloat.h"
 #include "llvm/ADT/SmallVector.h"
@@ -66,11 +67,16 @@ unsigned MOSMCCodeEmitter::encodeImm(const MCInst &MI, unsigned OpNo,
 
   if (MO.isExpr()) {
     if (isa<MOSMCExpr>(MO.getExpr())) {
-      // If the expression is already a MOSMCExpr,
-      // we shouldn't perform any more fixups. Without this check, we would
-      // instead create a fixup to the symbol named 'lo8(symbol)' which
-      // is not correct.
-      return getExprOpValue(MO.getExpr(), Fixups, STI);
+      MOSMCExpr const *MOSExpr = cast<MOSMCExpr>(MO.getExpr());
+      MCFixupKind FixupKind = static_cast<MCFixupKind>( MOSExpr->getFixupKind() );
+      if (FixupKind != MOS::Fixups::Addr16_Low &&
+          FixupKind != MOS::Fixups::Addr16_High) {
+        // If the expression is already a MOSMCExpr,
+        // we shouldn't perform any more fixups. Without this check, we would
+        // instead create a fixup to the symbol named 'lo8(symbol)' which
+        // is not correct.
+        return getExprOpValue(MO.getExpr(), Fixups, STI);
+      }
     }
 
     MCFixupKind FixupKind = static_cast<MCFixupKind>(Fixup); 
```

**Comment by laoo** [comment]
I've found that `MOSMCExpr` expression has separate path in `MOSMCCodeEmitter::encodeImm` method. I tried trivially rerouting `Addr16_Low` and `Addr16_High` expressions but it produced `Imm8` fixup. So the solution should be something more elaborate.

**Comment by johnwbyrd** [comment]
And how do you assign the abs variable?  Let's see a complete .s file that manifests the problem, please?

**Comment by laoo** [comment]
The problem emerges in the original example:
https://github.com/johnwbyrd/llvm-mos/blob/e9f9a775899a920403280186c796be757a0cff8c/llvm/test/MC/MOS/low-high-bytes.s#L10-L13

**Comment by johnwbyrd** [comment]
What do you perceive to be wrong with that code?

**Comment by laoo** [comment]
Nothing. The problem is in the assembler. Try it yourself:

```
d:\home\!llvm-mos-bin\bin>llvm-mc.exe -triple mos --filetype=obj -o=%TEMP%out.o ..\..\llvm-mos\llvm\test\MC\MOS\low-high-bytes.s

d:\home\!llvm-mos-bin\bin>llvm-objdump.exe -x %TEMP%out.o

b:\Temp\out.o:  file format elf32-mos

architecture: mos
start address: 0x00000000

Program Header:

Dynamic Section:
Sections:
Idx Name          Size     VMA      Type
  0               00000000 00000000
  1 .strtab       0000002f 00000000
  2 .text         00000011 00000000 TEXT
  3 .rela.text    00000018 00000000
  4 .symtab       00000050 00000000

SYMBOL TABLE:
00000000 l       .text  00000000 _start
0000ffd2 l       *ABS*  00000000 chrout
00000010 l       .text  00000000 done
00000000 l    d  .text  00000000 .text
RELOCATION RECORDS FOR [.text]:
OFFSET   TYPE                     VALUE
00000000 R_MOS_ADDR16_LO          .text
00000004 R_MOS_ADDR16_HI          .text


d:\home\!llvm-mos-bin\bin>llvm-objdump.exe -d %TEMP%out.o

b:\Temp\out.o:  file format elf32-mos


Disassembly of section .text:

00000000 <_start>:
       0: 00                            brk
       1: 00                            brk
       2: 86 10                         stx     16
       4: 00                            brk
       5: 00                            brk
       6: 85 11                         sta     17
       8: a2 d2                         ldx     #210
       a: 86 12                         stx     18
       c: a9 ff                         lda     #255
       e: 85 13                         sta     19

00000010 <done>:
      10: 60                            rts
```

Do you see that the relocation offsets are bad (off by one byte) and the disassembly shows that the generated code is broken?

**Comment by johnwbyrd** [comment]
Hmmm, this is interesting.  It does look to me like the code has an off by one error.

The culprit is likely in tryParseRelocExpression() in llvm/lib/Target/MOS/AsmParser/MOSAsmParser.cpp, which has a bunch of new code I just added for this specific functionality.  You can take a look at this sooner, or I can take a look in the next couple of days.

At some point Real Soon Now, I will make the < and > operators be shorthand for the more verbose versions you are using here.  Most 6502 compilers know this shorthand.

On an unrelated topic, it's good to see that you've got the tools compiling and running (more or less).

**Comment by laoo** [comment]
I've managed to patch the issue in `MOSMCCodeEmitter`:

```patch
diff --git a/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.cpp b/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.cpp
index 9117064c162..1132602a9d3 100644
--- a/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.cpp
+++ b/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.cpp
@@ -70,7 +71,7 @@ unsigned MOSMCCodeEmitter::encodeImm(const MCInst &MI, unsigned OpNo,
       // we shouldn't perform any more fixups. Without this check, we would
       // instead create a fixup to the symbol named 'lo8(symbol)' which
       // is not correct.
-      return getExprOpValue(MO.getExpr(), Fixups, STI);
+      return getExprOpValue(MO.getExpr(), Fixups, STI, Offset);
     }
 
     MCFixupKind FixupKind = static_cast<MCFixupKind>(Fixup);
@@ -85,7 +86,8 @@ unsigned MOSMCCodeEmitter::encodeImm(const MCInst &MI, unsigned OpNo,
 
 unsigned MOSMCCodeEmitter::getExprOpValue(const MCExpr *Expr,
                                           SmallVectorImpl<MCFixup> &Fixups,
-                                          const MCSubtargetInfo &STI) const {
+                                          const MCSubtargetInfo &STI,
+                                          unsigned int off ) const {
 
   MCExpr::ExprKind Kind = Expr->getKind();
 
@@ -102,7 +104,7 @@ unsigned MOSMCCodeEmitter::getExprOpValue(const MCExpr *Expr,
     }
 
     MCFixupKind FixupKind = static_cast<MCFixupKind>(MOSExpr->getFixupKind());
-    Fixups.push_back(MCFixup::create(0, MOSExpr, FixupKind));
+    Fixups.push_back(MCFixup::create(off, MOSExpr, FixupKind));
     return 0;
   }
 
diff --git a/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.h b/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.h
index 75e61c575d3..f62a0370ce9 100644
--- a/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.h
+++ b/llvm/lib/Target/MOS/MCTargetDesc/MOSMCCodeEmitter.h
@@ -86,7 +86,7 @@ private:
                                  const MCSubtargetInfo &STI) const;
 
   unsigned getExprOpValue(const MCExpr *Expr, SmallVectorImpl<MCFixup> &Fixups,
-                          const MCSubtargetInfo &STI) const;
+                          const MCSubtargetInfo &STI, unsigned int off ) const;
 
   /// Returns the binary encoding of operand.
   ///
```
It works in scenarios where I'm using these operators.

BTW Regarding the < and > shorthands and `tryParseRelocExpression` I was toying with the sources found that it _can_ be implemented in that function (I don't think that lexer should emit separate tokens for that):

```patch
diff --git a/llvm/lib/Target/MOS/AsmParser/MOSAsmParser.cpp b/llvm/lib/Target/MOS/AsmParser/MOSAsmParser.cpp
index fd587e0481d..633f650cbda 100644
--- a/llvm/lib/Target/MOS/AsmParser/MOSAsmParser.cpp
+++ b/llvm/lib/Target/MOS/AsmParser/MOSAsmParser.cpp
@@ -364,6 +364,28 @@ public:
       }
     }
 
+    MCExpr const *InnerExpression;
+    if (Parser.getTok().getKind() == AsmToken::Less ||
+        Parser.getTok().getKind() == AsmToken::Greater) {
+
+      switch ( Parser.getTok().getKind() )
+      {
+      case AsmToken::Less:
+        ModifierKind = MOSMCExpr::VK_MOS_ADDR16_LO;
+        break;
+      case AsmToken::Greater:
+        ModifierKind = MOSMCExpr::VK_MOS_ADDR16_HI;
+        break;
+      default:
+        assert(false);
+      }
+
+      Parser.Lex();
+    
+      if (getParser().parseExpression(InnerExpression))
+        return true;
+
+    } else {
       // Check if we have a target specific modifier (lo8, hi8, &c)
       if (Parser.getTok().getKind() != AsmToken::Identifier ||
           Parser.getLexer().peekTok().getKind() != AsmToken::LParen) {
@@ -395,7 +417,6 @@ public:
         Parser.Lex(); // Eat the sign and parenthesis
       }
 
-    MCExpr const *InnerExpression;
       if (getParser().parseExpression(InnerExpression))
         return true;
 
@@ -408,6 +429,7 @@ public:
       // If we have a modifier wrap the inner expression
       assert(Parser.getTok().getKind() == AsmToken::RParen);
       Parser.Lex(); // Eat closing parenthesis
+    }
 
     MCExpr const *Expression = MOSMCExpr::create(ModifierKind, InnerExpression,
                                                  IsNegated, getContext()); 
```

Regarding unrelated topic using the newest Visual Studio (16.7.3) it was sufficient to install Python and CMake and your command to build the solution worked. Only LLVM snapshot used in your project does not compile on that VS and I had to tweak it a little.

**Comment by johnwbyrd** [comment]
Cool, why don't you submit these diffs as pull requests and we'll do a code review on them together.

I've noticed something similar to your build problem over here... I suspect that rebasing my work on the latest LLVM will make that build problem go away... Could you please open a bug request for that too.

I appreciate you taking the time to make these changes in a coordinated manner :)  Let's make sure they don't get lost.

**Comment by laoo** [comment]
OK. My first pull-request ever. I hope I've done it correctly.
#3

**Comment by johnwbyrd** [comment]
Added some comments to the patch -- I'd love to see your test case actually check for correct and incorrect results.

**Comment by johnwbyrd** [comment]
Fixed with #3
